### PR TITLE
Implement RunnablePassthrough.assign(...)

### PR DESCRIPTION
**Pull request by nfcampos**
Passes through dict input and assigns additional keys

### CODE REVIEW

File: libs/langchain/langchain/schema/runnable/base.py [hunk 1]
```diff
@@ -1748,30 +1749,6 @@ async def input_aiter() -> AsyncIterator[Input]:
             yield chunk
 
 
-class RunnableMapChunk(Dict[str, Any]):
```
Comment (nfcampos) [inline libs/langchain/langchain/schema/runnable/base.py:1751]: Moved to utils

File: libs/langchain/langchain/schema/runnable/base.py [hunk 2]
```diff
@@ -1814,14 +1791,18 @@ def InputType(self) -> Any:
 
     @property
     def input_schema(self) -> type[BaseModel]:
-        if all(not s.input_schema.__custom_root_type__ for s in self.steps.values()):
+        if all(
```
Comment (nfcampos) [inline libs/langchain/langchain/schema/runnable/base.py:1794]: This works in more cases

**Comment by eyurtsev** [comment]
looking

File: libs/langchain/langchain/schema/runnable/passthrough.py [hunk 3]
```diff
@@ -38,6 +56,30 @@ def InputType(self) -> Any:
     def OutputType(self) -> Any:
         return self.input_type or Any
 
+    @classmethod
+    def assign(
```
Comment (eyurtsev) [inline libs/langchain/langchain/schema/runnable/passthrough.py:60]: Should we add an `assign` method to RunnableMap?

File: libs/langchain/tests/unit_tests/schema/runnable/test_runnable.py [hunk 4]
```diff
@@ -2016,6 +2017,104 @@ def test_deep_stream() -> None:
...
+    assert chain_with_assign.input_schema.schema() == {
+        "title": "PromptInput",
+        "type": "object",
+        "properties": {"question": {"title": "Question"}},
+    }
+    assert chain_with_assign.output_schema.schema() == {
+        "title": "RunnableAssignOutput",
+        "type": "object",
+        "properties": {
+            "str": {"title": "Str"},
```
Comment (eyurtsev) [inline libs/langchain/tests/unit_tests/schema/runnable/test_runnable.py:2051]: Is there a way to fetch the type for `str` here? or too hard?

**Review by eyurtsev — APPROVED** [review]

File: libs/langchain/langchain/schema/runnable/passthrough.py (hunk 3 above)
Comment (nfcampos) [reply libs/langchain/langchain/schema/runnable/passthrough.py:60]: What would we use that for?

File: libs/langchain/langchain/schema/runnable/passthrough.py [hunk 5]
```diff
@@ -65,3 +107,155 @@ async def atransform(
...
+
+    def invoke(
+        self,
+        input: Dict[str, Any],
+        config: Optional[RunnableConfig] = None,
+        **kwargs: Any,
+    ) -> Dict[str, Any]:
+        assert isinstance(input, dict)
+        return {
+            **input,
```
[the remark this answers is not in the document]
Comment (nfcampos) [reply libs/langchain/langchain/schema/runnable/passthrough.py:167]: No, this is on purpose, so you can eg modify an existing key

[the remark this answers is not in the document]
Comment (nfcampos) [reply libs/langchain/langchain/schema/runnable/passthrough.py:167]: Same semantics as JS Object.assign

File: libs/langchain/tests/unit_tests/schema/runnable/test_runnable.py (hunk 4 above)
Comment (nfcampos) [reply libs/langchain/tests/unit_tests/schema/runnable/test_runnable.py:2051]: No, this is being inferred by analysis of the `itemgetter`, which doesn't include typing information

**Pull request merged**